src/main/java/com/hivemq/bootstrap/ClientConnection.java

```diff
@@ -105,12 +102,159 @@ public class ClientConnection {
+
+        final ClientConnectionContext clientConnectionContext =
+                channel.attr(ClientConnectionContext.CHANNEL_ATTRIBUTE_NAME).get();
+
+        checkArgument(clientConnectionContext instanceof ClientConnection);
+
+        return (ClientConnection) clientConnectionContext;
+    }
+
+    public static @NotNull ClientConnection from(final @NotNull ClientConnectionContext clientConnectionContext) {
+        checkArgument(clientConnectionContext instanceof UndefinedClientConnection);
+
+        final UndefinedClientConnection context = (UndefinedClientConnection) clientConnectionContext;
+
+        checkNotNull(context.clientId, "Client id must not be null.");
+        checkNotNull(context.clientState, "Client state must not be null.");
+        checkNotNull(context.protocolVersion, "Protocol version must not be null.");
+        checkNotNull(context.connectedListener, "Connected listener must not be null.");
+
+        final ClientConnection clientConnection = new ClientConnection(
+                context.channel,
+                context.publishFlushHandler,
+                context.clientState,
+                context.protocolVersion,
+                context.clientId,
+                context.cleanStart,
+                context.authPermissions,
+                context.connectedListener,
+                context.connectMessage,
+                context.clientReceiveMaximum,
+                context.connectKeepAlive,
+                context.queueSizeMaximum,
+                context.clientSessionExpiryInterval,
+                context.connectReceivedTimestamp,
+                context.topicAliasMapping,
+                context.clientIdAssigned,
+                context.incomingPublishesSkipRest,
+                context.requestResponseInformation,
+                context.requestProblemInformation,
+                context.disconnectFuture,
+                context.connectionAttributes,
+                context.sendWill,
+                context.preventLwt,
+                context.authCertificate,
+                context.authSniHostname,
+                context.authCipherSuite,
+                context.authProtocol,
+                context.authUsername,
+                context.authPassword,
+                context.authConnect,
+                context.authMethod,
+                context.authData,
+                context.authUserProperties,
+                context.authFuture,
+                context.maxPacketSizeSend,
+                context.extensionClientContext,
+                context.extensionClientEventListeners,
+                context.extensionClientAuthenticators,
+                context.extensionClientAuthorizers,
+                context.extensionClientInformation,
+                context.extensionConnectionInformation);
+
+        context.getChannel().attr(ClientConnectionContext.CHANNEL_ATTRIBUTE_NAME).set(clientConnection);
+        return clientConnection;
+    }
+
+    public ClientConnection(
```

Comment:
I think this can be private if tests use a mock.

Reply:
As the ClientConnection is an essential part of every connection established in the unit and integration test it would require to touch many test in this regard. Would prefer to let it stay public.